File: codegen/cpp/dual.cpp
// 3D Projective Geometric Algebra
// Written by a generator written by enki.
#include <stdio.h>
#include <cmath>
#include <array>

#define PI 3.14159265358979323846

static const char* basis[] = { "1","e0" };

class DUAL {
  public:
    DUAL ()  { std::fill( mvec, mvec + sizeof( mvec )/4, 0.0f ); }
    DUAL (float f, int idx=0) { std::fill( mvec, mvec + sizeof( mvec )/4, 0.0f ); mvec[idx] = f; }
    float& operator [] (size_t idx) { return mvec[idx]; }
    const float& operator [] (size_t idx) const { return mvec[idx]; }
    DUAL log () { int n=0; for (int i=0,j=0;i<2;i++) if (mvec[i]!=0.0f) { n++; printf("%s%0.7g%s",(j>0)?" + ":"",mvec[i],(i==0)?"":basis[i]); j++; };if (n==0) printf("0");  printf("\n"); return *this; }
    DUAL Conjugate(); 
    DUAL Involute();
    float norm();
    float inorm();
    DUAL normalized();
  private:  
    float mvec[2];
};


//***********************
// DUAL.Reverse : res = ~a
// Reverse the order of the basis blades.
//***********************
inline DUAL operator ~ (const DUAL &a) {
  DUAL res;
  res[0]=a[0];
  res[1]=a[1];
  return res;
};

//***********************
// DUAL.Dual : res = !a
// Poincare duality operator.
//***********************
inline DUAL operator ! (const DUAL &a) {
  DUAL res;
  res[0]=a[1];
  res[1]=a[0];
  return res;
};

//***********************
// DUAL.Conjugate : res = a.Conjugate()
// Clifford Conjugation
//***********************
inline DUAL DUAL::Conjugate () {
  DUAL res;
  res[0]=this->mvec[0];
  res[1]=-this->mvec[1];
  return res;
};

//***********************
// DUAL.Involute : res = a.Involute()
// Main involution
//***********************
inline DUAL DUAL::Involute () {
  DUAL res;
  res[0]=this->mvec[0];
  res[1]=-this->mvec[1];
  return res;
};

//***********************
// DUAL.Mul : res = a * b 
// The geometric product.
//***********************
inline DUAL operator * (const DUAL &a, const DUAL &b) {
  DUAL res;
  res[0]=b[0]*a[0];
  res[1]=b[1]*a[0]+b[0]*a[1];
  return res;
};

//***********************
// DUAL.Wedge : res = a ^ b 
// The outer product. (MEET)
//***********************
inline DUAL operator ^ (const DUAL &a, const DUAL &b) {
  DUAL res;
  res[0]=b[0]*a[0];
  res[1]=b[1]*a[0]+b[0]*a[1];
  return res;
};

//***********************
// DUAL.Vee : res = a & b 
// The regressive product. (JOIN)
//***********************
inline DUAL operator & (const DUAL &a, const DUAL &b) {
  DUAL res;
  res[1]=b[1]*a[1];
  res[0]=b[0]*a[1]+b[1]*a[0];
  return res;
};

//***********************
// DUAL.Dot : res = a | b 
// The inner product.
//***********************
inline DUAL operator | (const DUAL &a, const DUAL &b) {
  DUAL res;
  res[0]=b[0]*a[0];
  res[1]=b[1]*a[0]+b[0]*a[1];
  return res;
};

//***********************
// DUAL.Add : res = a + b 
// Multivector addition
//***********************
inline DUAL operator + (const DUAL &a, const DUAL &b) {
  DUAL res;
      res[0] = a[0]+b[0];
    res[1] = a[1]+b[1];
  return res;
};

//***********************
// DUAL.Sub : res = a - b 
// Multivector subtraction
//***********************
inline DUAL operator - (const DUAL &a, const DUAL &b) {
  DUAL res;
      res[0] = a[0]-b[0];
    res[1] = a[1]-b[1];
  return res;
};

//***********************
// DUAL.smul : res = a * b 
// scalar/multivector multiplication
//***********************
inline DUAL operator * (const float &a, const DUAL &b) {
  DUAL res;
      res[0] = a*b[0];
    res[1] = a*b[1];
  return res;
};

//***********************
// DUAL.muls : res = a * b 
// multivector/scalar multiplication
//***********************
inline DUAL operator * (const DUAL &a, const float &b) {
  DUAL res;
      res[0] = a[0]*b;
    res[1] = a[1]*b;
  return res;
};

//***********************
// DUAL.sadd : res = a + b 
// scalar/multivector addition
//***********************
inline DUAL operator + (const float &a, const DUAL &b) {
  DUAL res;
    res[0] = a+b[0];
      res[1] = b[1];
  return res;
};

//***********************
// DUAL.adds : res = a + b 
// multivector/scalar addition
//***********************
inline DUAL operator + (const DUAL &a, const float &b) {
  DUAL res;
    res[0] = a[0]+b;
      res[1] = a[1];
  return res;
};


inline float DUAL::norm() { return sqrt(abs(((*this)*Conjugate()).mvec[0])); }
inline float DUAL::inorm() { return (!(*this)).norm(); }
inline DUAL DUAL::normalized() { return (*this) * (1/norm()); }


static DUAL e0(1.0f,1);


int main (int argc, char **argv) {
  
  printf("e0*e0         : "); (e0*e0).log();
  printf("pss           : "); e0.log();
  printf("pss*pss       : "); (e0*e0).log();

  return 0;
}

File: codegen/rust/dual.rs
// 3D Projective Geometric Algebra
// Written by a generator written by enki.
#![allow(unused_imports)]
#![allow(dead_code)]
#![allow(non_upper_case_globals)]
#![allow(non_snake_case)]
#![allow(non_camel_case_types)]
//#![feature(const_slice_len)]

use std::fmt;
use std::ops::{Index,IndexMut,Add,Sub,Mul,BitAnd,BitOr,BitXor,Not};

type float_t = f64;

// use std::f64::consts::PI;
const PI: float_t = 3.14159265358979323846;

const basis: &'static [&'static str] = &[ "1","e0" ];
const basis_count: usize = basis.len();

#[derive(Default,Debug,Clone,PartialEq)]
struct DUAL {
    mvec: Vec<float_t>
}

impl DUAL {
    pub fn zero() -> Self {
        Self {
            mvec: vec![0.0; basis_count]
        }
    }

    pub fn new(f: float_t, idx: usize) -> Self {
        let mut ret = Self::zero();
        ret.mvec[idx] = f;
        ret
    }

    // basis vectors are available as methods
    pub fn e0() -> Self { DUAL::new(1.0, 1) }
}

impl Index<usize> for DUAL {
    type Output = float_t;

    fn index<'a>(&'a self, index: usize) -> &'a Self::Output {
        &self.mvec[index]
    }
}

impl IndexMut<usize> for DUAL {
    fn index_mut<'a>(&'a mut self, index: usize) -> &'a mut Self::Output {
        &mut self.mvec[index]
    }
}

impl fmt::Display for DUAL {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let mut n = 0;
        let ret = self.mvec.iter().enumerate().filter_map(|(i, &coeff)| {
            if coeff > 0.00001 || coeff < -0.00001 {
                n = 1;
                Some(format!("{}{}", 
                        format!("{:.*}", 7, coeff).trim_end_matches('0').trim_end_matches('.'),
                        if i > 0 { basis[i] } else { "" }
                    )
                )
            } else {
                None
            }
        }).collect::<Vec<String>>().join(" + ");
        if n==0 { write!(f,"0") } else { write!(f, "{}", ret) }
    }
}

macro_rules! define_binary_op(
    (
        // Operator, operator method, and scalar bounds.
        $Op: ident, $op: ident;
        // Argument identifiers and types + output.
        $lhs: ident: $Lhs: ty, $rhs: ident: $Rhs: ty, Output = $Result: ty;
        // Operator actual implementation.
        $action: expr;
        // Lifetime.
        $($lives: tt),*
    ) => {
       impl<$($lives ,)*> $Op<$Rhs> for $Lhs {
           type Output = $Result;

           #[inline]
           fn $op($lhs, $rhs: $Rhs) -> Self::Output {
               $action
           }
       }
    }
);

macro_rules! define_binary_op_all(
    (
        // Operator, operator method, and scalar bounds.
        $Op: ident, $op: ident;
        // Argument identifiers and types + output.
        $lhs: ident: $Lhs: ty, $rhs: ident: $Rhs: ty, Output = $Result: ty;
        // Operators actual implementations.
        [val val] => $action_val_val: expr;
        [ref val] => $action_ref_val: expr;
        [val ref] => $action_val_ref: expr;
        [ref ref] => $action_ref_ref: expr;
    ) => {
        define_binary_op!(
            $Op, $op;
            $lhs: $Lhs, $rhs: $Rhs, Output = $Result;
            $action_val_val;
        );

        define_binary_op!(
            $Op, $op;
            $lhs: &'a $Lhs, $rhs: $Rhs, Output = $Result;
            $action_ref_val;
            'a
        );

        define_binary_op!(
            $Op, $op;
            $lhs: $Lhs, $rhs: &'b $Rhs, Output = $Result;
            $action_val_ref;
            'b
        );

        define_binary_op!(
            $Op, $op;
            $lhs: &'a $Lhs, $rhs: &'b $Rhs, Output = $Result;
            $action_ref_ref;
            'a, 'b
        );
    }
);

// TODO define_unary_op



// Reverse
// Reverse the order of the basis blades.
impl DUAL {
    pub fn Reverse(self: & Self) -> DUAL {
        let mut res = DUAL::zero();
        let a = self;
        res[0]=a[0];
        res[1]=a[1];
        res
    }
}

// Dual
// Poincare duality operator.
impl DUAL {
    pub fn Dual(self: & Self) -> DUAL {
        let mut res = DUAL::zero();
        let a = self;
        res[0]=a[1];
        res[1]=a[0];
        res
    }
}

impl Not for & DUAL {
    type Output = DUAL;

    fn not(self: Self) -> DUAL {
        let mut res = DUAL::zero();
        let a = self;
        res[0]=a[1];
        res[1]=a[0];
        res
    }
}

// Conjugate
// Clifford Conjugation
impl DUAL {
    pub fn Conjugate(self: & Self) -> DUAL {
        let mut res = DUAL::zero();
        let a = self;
        res[0]=a[0];
        res[1]=-a[1];
        res
    }
}

// Involute
// Main involution
impl DUAL {
    pub fn Involute(self: & Self) -> DUAL {
        let mut res = DUAL::zero();
        let a = self;
        res[0]=a[0];
        res[1]=-a[1];
        res
    }
}

// Mul
// The geometric product.

define_binary_op_all!(
    Mul,
    mul;
    self: DUAL, b: DUAL, Output = DUAL;
    [val val] => &self * &b;
    [ref val] =>  self * &b;
    [val ref] => &self *  b;
    [ref ref] => {
        let mut res = DUAL::zero();
        let a = self;
        res[0]=b[0]*a[0];
		res[1]=b[1]*a[0]+b[0]*a[1];
        res
    };
);


// Wedge
// The outer product. (MEET)

define_binary_op_all!(
    BitXor,
    bitxor;
    self: DUAL, b: DUAL, Output = DUAL;
    [val val] => &self ^ &b;
    [ref val] =>  self ^ &b;
    [val ref] => &self ^  b;
    [ref ref] => {
        let mut res = DUAL::zero();
        let a = self;
        res[0]=b[0]*a[0];
		res[1]=b[1]*a[0]+b[0]*a[1];
        res
    };
);


// Vee
// The regressive product. (JOIN)

define_binary_op_all!(
    BitAnd,
    bitand;
    self: DUAL, b: DUAL, Output = DUAL;
    [val val] => &self & &b;
    [ref val] =>  self & &b;
    [val ref] => &self &  b;
    [ref ref] => {
        let mut res = DUAL::zero();
        let a = self;
        res[1]=(a[1]*b[1]);
		res[0]=(a[0]*b[1]+a[1]*b[0]);
        res
    };
);


// Dot
// The inner product.

define_binary_op_all!(
    BitOr,
    bitor;
    self: DUAL, b: DUAL, Output = DUAL;
    [val val] => &self | &b;
    [ref val] =>  self | &b;
    [val ref] => &self |  b;
    [ref ref] => {
        let mut res = DUAL::zero();
        let a = self;
        res[0]=b[0]*a[0];
		res[1]=b[1]*a[0]+b[0]*a[1];
        res
    };
);


// Add
// Multivector addition

define_binary_op_all!(
    Add,
    add;
    self: DUAL, b: DUAL, Output = DUAL;
    [val val] => &self + &b;
    [ref val] =>  self + &b;
    [val ref] => &self +  b;
    [ref ref] => {
        let mut res = DUAL::zero();
        let a = self;
        res[0] = a[0]+b[0];
		res[1] = a[1]+b[1];
        res
    };
);


// Sub
// Multivector subtraction

define_binary_op_all!(
    Sub,
    sub;
    self: DUAL, b: DUAL, Output = DUAL;
    [val val] => &self - &b;
    [ref val] =>  self - &b;
    [val ref] => &self -  b;
    [ref ref] => {
        let mut res = DUAL::zero();
        let a = self;
        res[0] = a[0]-b[0];
		res[1] = a[1]-b[1];
        res
    };
);


// smul
// scalar/multivector multiplication

define_binary_op_all!(
    Mul,
    mul;
    self: float_t, b: DUAL, Output = DUAL;
    [val val] => &self * &b;
    [ref val] =>  self * &b;
    [val ref] => &self *  b;
    [ref ref] => {
        let mut res = DUAL::zero();
        let a = self;
        res[0] = a*b[0];
        res[1] = a*b[1];
        res
    };
);


// muls
// multivector/scalar multiplication

define_binary_op_all!(
    Mul,
    mul;
    self: DUAL, b: float_t, Output = DUAL;
    [val val] => &self * &b;
    [ref val] =>  self * &b;
    [val ref] => &self *  b;
    [ref ref] => {
        let mut res = DUAL::zero();
        let a = self;
        res[0] = a[0]*b;
        res[1] = a[1]*b;
        res
    };
);


// sadd
// scalar/multivector addition

define_binary_op_all!(
    Add,
    add;
    self: float_t, b: DUAL, Output = DUAL;
    [val val] => &self + &b;
    [ref val] =>  self + &b;
    [val ref] => &self +  b;
    [ref ref] => {
        let mut res = DUAL::zero();
        let a = self;
        res[0] = a+b[0];
        res[1] = b[1];
        res
    };
);


// adds
// multivector/scalar addition

define_binary_op_all!(
    Add,
    add;
    self: DUAL, b: float_t, Output = DUAL;
    [val val] => &self + &b;
    [ref val] =>  self + &b;
    [val ref] => &self +  b;
    [ref ref] => {
        let mut res = DUAL::zero();
        let a = self;
        res[0] = a[0]+b;
        res[1] = a[1];
        res
    };
);


// ssub
// scalar/multivector subtraction

define_binary_op_all!(
    Sub,
    sub;
    self: float_t, b: DUAL, Output = DUAL;
    [val val] => &self - &b;
    [ref val] =>  self - &b;
    [val ref] => &self -  b;
    [ref ref] => {
        let mut res = DUAL::zero();
        let a = self;
        res[0] = a-b[0];
        res[1] = -b[1];
        res
    };
);


// subs
// multivector/scalar subtraction

define_binary_op_all!(
    Sub,
    sub;
    self: DUAL, b: float_t, Output = DUAL;
    [val val] => &self - &b;
    [ref val] =>  self - &b;
    [val ref] => &self -  b;
    [ref ref] => {
        let mut res = DUAL::zero();
        let a = self;
        res[0] = a[0]-b;
        res[1] = a[1];
        res
    };
);


impl DUAL {
    pub fn norm(self: & Self) -> float_t {
        let scalar_part = (self * self.Conjugate())[0];

        scalar_part.abs().sqrt()
    }

    pub fn inorm(self: & Self) -> float_t {
        self.Dual().norm()
    }

    pub fn normalized(self: & Self) -> Self {
        self * (1.0 / self.norm())
    }
    
    

}


fn main() {

  println!("e0*e0         : {}", DUAL::e0() * DUAL::e0());
  println!("pss           : {}", DUAL::e0());
  println!("pss*pss       : {}", DUAL::e0() * DUAL::e0());

}